src/H5Iint.c

```diff
@@ -8727,24 +8791,17 @@ H5I__discard_mt_id_info(H5I_mt_id_info_t * id_info_ptr)
     result = atomic_compare_exchange_strong(&(id_info_ptr->on_fl), &on_fl, TRUE);
     assert( result );
 
-#if 1 /* JRM */
-    result = atomic_compare_exchange_strong(&(id_info_ptr->re_allocable), &on_fl, TRUE);
-    assert( result );
-#endif /* JRM */
+    atomic_store(&(id_info_ptr->serial_num), atomic_fetch_add(&(H5I_mt_g.id_next_sn), 1ULL));
```

Comment:
If `H5I__discard_mt_id_info()` can be used concurrently, I don't think this is guaranteed to assign a unique SN to each free list entry. 

1. Thread 1 begin `H5T__discard_mt_id_info()`, executes up to `atomic_fetch_add()`, but is pre-empted before writing out the SN to the new FL node.
2. Thread 2 begins `H5T__discard_mt_id_info()` on a different ID info and also executes up to `atomic_fetch_add()`. The previous value of `id_next_sn` is skipped and never written to a free list node.
3. Now, regardless of whether Thread 1 or 2 moves on to the `atomic_store()` first, they will both write the same serial number.

Reply:
I originally had those atomic functions separated out instead of combined, and when John was cleaning up the code he combined them. I did not pick up on the potential concurrency problem when I looked at and I don't know if John thought about that either. I will bring this up with him. Thank you catching this and for the valuable insight.

Reply:
This code can certainly assign serial numbers out of order in the free list, but if atomic_fetch_add() behaves as documented in the C11 standard, I don't see how we skip a serial number or use a serial number twice.  As I understand it, the whole point of atomics is that they are always serialized one way or the other.  Thus both threads will execute atomic_fetch_add() and get unique serial numbers.

That said, I think the point is moot from the perspective of managing the free list -- which is sort of the point of the new algorithm.

Reply:
I think you're right that serial numbers can't be skipped or double added - in the example I laid out, Thread 1 and Thread 2 would still write out different serial numbers, because the second execution of `atomic_fetch_add()` wouldn't modify the return value of the first execution.

As for out of order serial number assignment, since the algorithm uses a max-reallocatable-SN to decide when entries can be re-allocated, I think you're right that it's still safe, since an out-of-order assignment would at worst result in some list entries not being re-allocated for an unusually long time, which should be fine.